[18.09] Fix pretty printing of shed_tool_conf.xml file

mvdbeek opened the pull request:
We rewrite the shed_tool_conf.xml file during tool installations.
Lately I've seen entries like the following:

```
<section id="textutil" name="Text Manipulation" version="">\\\\n
<tool file="toolshed.g2.bx.psu.edu/repos/mvdbeek/add_input_name_as_column/b31219f26a8f/add_input_name_as_column/add_input_name_as_column.xml" guid="toolshed.g2.bx.psu.edu/repos/mvdbeek/add_input_name_as_column/addName/0.1.1">\\\\n      <tool_shed>toolshed.g2.bx.psu.edu</tool_shed>\\\\n        <repository_name>add_input_name_as_column</repository_name>\\\\n        <repository_owner>mvdbeek</repository_owner>\\\\n        <installed_changeset_revision>b31219f26a8f</installed_changeset_revision>\\\\n        <id>toolshed.g2.bx.psu.edu/repos/mvdbeek/add_input_name_as_column/addName/0.1.1</id>\\\\n        <version>0.1.1</version>\\\\n    </tool>\\\\n</section>\\\\n\\\'\\nb\\\'\'\nb\''
b'






<section id="features" name="Extract Features" version="">\\\\n  <tool file="toolshed.g2.bx.psu.edu/repos/iuc/ucsc_fasplit/dc36d50254d8/ucsc_fasplit/fasplit.xml" guid="toolshed.g2.bx.psu.edu/repos/iuc/ucsc_fasplit/fasplit/332">\\\\n      <tool_shed>toolshed.g2.bx.psu.edu</tool_shed>\\\\n        <repository_name>ucsc_fasplit</repository_name>\\\\n        <repository_owner>iuc</repository_owner>\\\\n        <installed_changeset_revision>dc36d50254d8</installed_changeset_revision>\\\\n        <id>toolshed.g2.bx.psu.edu/repos/iuc/ucsc_fasplit/fasplit/332</id>\\\\n        <version>332</version>\\\\n    </tool>\\\\n</section>\\\\n\\\'\\nb\\\'\'\nb\''
b'





```

This fixes that issue.

This drops xml_to_string from lib/tool_shed/util/xml_util.py
and uses xml_to_string form lib/galaxy/utils/__init__.py.
Also reduces plain string manipulation of xml structures.
Uses xml.dom.minidom to make xml pretty if necessary.

mvdbeek committed:
Fix messed up printing of shed_tool_conf.xml file

This drops xml_to_string from lib/tool_shed/util/xml_util.py
and uses xml_to_string form lib/galaxy/utils/__init__.py.
Also reduces plain string manipulation of xml structures.
Uses xml.dom.minidom to make xml pretty if necessary.

nsoranzo commented on lib/galaxy/util/__init__.py:
```diff
@@ -235,19 +236,10 @@ def parse_xml_string(xml_string):
-        else:
-            raise e
+        pretty_string = xml.dom.minidom.parseString(xml_str).toprettyxml(indent='    ')
+        return "\n".join([line for line in pretty_string.split('\n') if not re.match(r'^[\s\\nb\']*$', line)])
```
This method now does nothing if `pretty` is `False.

mvdbeek replied:
Oops, thanks, I fixed that in a rebase

mvdbeek committed:
Fix toolshed tests

mvdbeek committed:
Drop threading lock and global try/except

The Lock wasn't actually preventing anything, and I don't think
anything would need ot be protected here. mkdtemp should be thread-safe.

We also remove the try/except Statement. I think it'd be better to
expose any exceptions here instead of producing broken capsules and
logging an exception.

mvdbeek committed:
Fix capsule export in docker testing

Found this by removing the blank try/except in the previous commit.
This only affects testing when python can't determine the correct
locale and falls back to ASCII.

mvdbeek commented:
@galaxybot test this

mvdbeek commented:
@galaxybot test py3

nsoranzo commented on lib/tool_shed/galaxy_install/tools/data_manager.py:
```diff
@@ -29,21 +30,18 @@ def data_manager_config_elems_to_xml_file(self, config_elems, config_filename):
-            lock.release()
+                fh.write(xml_to_string(root))
+        except Exception as e:
+            log.exception("Exception in DataManagerHandler.data_manager_config_elems_to_xml_file:\n %s", str(e))
```
This change was not needed, `log.exception()` already adds the exception info to the log message.

nsoranzo commented on lib/tool_shed/util/xml_util.py:
```diff
@@ -20,17 +22,13 @@ def comment(self, data):
+    tmp_str = xml_to_string(elem, pretty=True)
+    fh = tempfile.NamedTemporaryFile(prefix="tmp-toolshed-cawrf", delete=False)
     tmp_filename = fh.name
     fh.close()
```
You can use a `with` statement here too.
